### PR TITLE
Deprecate find()

### LINKED ISSUE #1694: Deprecate find()
`find(strategy, condition)` can be replaced with `strategy.filter(condition).example()`.  It does some other stuff too, but (a) `find()` is actually not very useful, and (b) it's a messy and these days second-class execution engine compared to `@given` - see #1609 and in particular #1213.

This is basically a straight clean-up exercise to reduce the API surface that we have to maintain, and ensure that each piece is orthogonal.

### PR DESCRIPTION
**Pull request by Zac-HD**
This PR *finally* closes #1694, after other PRs ported our internal uses of `find()` over to `@given`.

In particular, this consists mostly of documentation changes and our `@checks_deprecated_behaviour` test helper - there were only a few tests left using `find()` without specifically existing *to* test `find()`.  And one important call to `note_deprecation`, of course!

### CODE REVIEW

File: hypothesis-python/docs/data.rst [hunk 1]
```diff
@@ -151,10 +151,6 @@ Most of the time you probably don't want ``flatmap``, but unlike ``filter`` and
 ``flatmap`` allows genuinely new data generation that you wouldn't otherwise be
 able to easily do.
 
-(If you know Haskell: Yes, this is more or less a monadic bind. If you don't
```
Comment (DRMacIver) [inline hypothesis-python/docs/data.rst]: This seems an unrelated removal (and also I'd like to keep this line)

File: hypothesis-python/src/hypothesis/core.py [hunk 2]
```diff
@@ -1054,6 +1055,11 @@ def find(
     # type: (...) -> Any
     """Returns the minimal example from the given strategy ``specifier`` that
     matches the predicate function ``condition``."""
+    note_deprecation(
+        "`find(s, f)` is deprecated in favor of `s.filter(f).example()`, which "
```
Comment (DRMacIver) [inline hypothesis-python/src/hypothesis/core.py]: I don't think this deprecation notice is really true. `find` is being deprecated in favour of not using this functionality because basically nobody does and we don't want to support it. `@given` has no real shrinking advantage over `find`, and I don't especially want to encourage people to use `.example()` _more_ than they already do.

**Comment by Zac-HD** [comment]
CC @HypothesisWorks/hypothesis-python-contributors for review?
